cmd/clusterctl/client/tree/discovery.go

```diff
@@ -105,25 +122,98 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
-		return tree, nil
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
-	tree.Add(cluster, workers)
 
+	if len(machinesList.Items) != len(controlPlaneMachines) {
+		tree.Add(cluster, workers)
+		err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+		if err != nil {
+			return nil, err
+		}
+
+		// Handles orphan machines.
+		if len(machineMap) < len(machinesList.Items) {
+			other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+			tree.Add(workers, other)
+
+			for i := range machinesList.Items {
+				m := &machinesList.Items[i]
+				if _, ok := machineMap[m.Name]; ok {
+					continue
+				}
+				addMachineFunc(other, m)
+			}
+		}
+	} else if len(machinePoolList.Items) > 0 {
+		tree.Add(cluster, workers)
+		addMachinePoolsToObjectTree(workers, machinePoolList, tree)
+	}
+
+	return tree, nil
+}
+
+func addClusterResourceSetsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBinding, err := getResourceSetBindingInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		resourceSetGroup := VirtualObject(cluster.Namespace, "ClusterResourceSetGroup", "ClusterResourceSets")
+		tree.Add(cluster, resourceSetGroup)
+
+		for _, binding := range resourceSetBinding.Spec.Bindings {
+			resourceSetRefObject := ObjectReferenceObject(&corev1.ObjectReference{
+				Kind:       "ClusterResourceSet", // TODO: does this kind name exist anywhere as a const?
+				Namespace:  cluster.Namespace,
+				Name:       binding.ClusterResourceSetName,
+				APIVersion: addonsv1.GroupVersion.String(),
+			})
+			tree.Add(resourceSetGroup, resourceSetRefObject, ObjectMetaName("ClusterResourceSet"))
+		}
+	}
+}
+
+func addControlPlane(cluster *clusterv1.Cluster, controlPlane *unstructured.Unstructured, tree *ObjectTree, options DiscoverOptions) {
+	tree.Add(cluster, controlPlane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+
+	if options.ShowTemplates {
+		// Add control plane infrastructure ref using spec fields guaranteed in contract
+		infrastructureRef, found, err := unstructured.NestedMap(controlPlane.UnstructuredContent(), "spec", "machineTemplate", "infrastructureRef")
+		if err == nil && found {
+			infrastructureObjectRef := &corev1.ObjectReference{
+				Kind:       infrastructureRef["kind"].(string),
+				Namespace:  infrastructureRef["namespace"].(string),
+				Name:       infrastructureRef["name"].(string),
+				APIVersion: infrastructureRef["apiVersion"].(string),
+			}
+
+			machineTemplateRefObject := ObjectReferenceObject(infrastructureObjectRef)
+			var templateParent client.Object
+			if options.AddTemplateVirtualNode {
+				templateParent = VirtualObject(cluster.Namespace, "TemplateGroup", "Templates")
```

Comment:
Should this have zOrder -1 so it goes before everything the machines (same for other templates nodes)